codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,457 @@
+            try {
+                http2Handler.connection().forEachActiveStream(new Http2StreamVisitor() {
+                    @Override
+                    public boolean visit(Http2Stream stream) {
+                        if (stream.id() > lastStreamId
+                                && http2Handler.connection().local().isValidStreamId(stream.id())) {
+                            StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+                            if (streamInfo != null) {
+                                streamInfo.childChannel.pipeline()
+                                    .fireUserEventTriggered(goAway.retain());
+                            }
+                        }
+                        return true;
+                    }
+                });
+            } catch (Throwable t) {
+                ctx.invoker().invokeExceptionCaught(ctx, t);
+            }
+            ctx.fireUserEventTriggered(goAway.retain());
+        }
+    }
+
+    class InternalHttp2ConnectionHandler extends Http2ConnectionHandler {
+        public InternalHttp2ConnectionHandler(Http2Connection connection,
+                Http2FrameReader frameReader, Http2FrameWriter frameWriter,
+                Http2FrameListener listener) {
+            super(connection, frameReader, frameWriter, listener);
+        }
+
+        @Override
+        protected void onStreamError(ChannelHandlerContext ctx, Throwable cause,
+                Http2Exception.StreamException http2Ex) {
+            try {
+                Http2Stream stream = http2Handler.connection().stream(http2Ex.streamId());
+                if (stream == null) {
+                    return;
+                }
+                StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+                if (streamInfo == null) {
+                    return;
+                }
+                streamInfo.childChannel.pipeline().fireExceptionCaught(http2Ex);
+            } finally {
+                super.onStreamError(ctx, cause, http2Ex);
+            }
+        }
+    }
+
+    class FrameListener extends Http2FrameAdapter {
+        @Override
+        public void onRstStreamRead(ChannelHandlerContext ctx, int streamId, long errorCode)
+                throws Http2Exception {
+            Http2Stream stream = http2Handler.connection().stream(streamId);
+            StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+            // Use a user event in order to circumvent read queue.
+            streamInfo.childChannel.pipeline().fireUserEventTriggered(new DefaultHttp2ResetFrame(errorCode));
+        }
+
+        @Override
+        public void onHeadersRead(ChannelHandlerContext ctx, int streamId,
+                Http2Headers headers, int streamDependency, short weight, boolean
+                exclusive, int padding, boolean endStream) throws Http2Exception {
+            onHeadersRead(ctx, streamId, headers, padding, endStream);
```

Comment:
The are intentionally left out in part because of the questions you are asking. I've not thought about them too much or have a concrete design in my head. But also, they are nowhere near the core use case or necessary to prove the general approach that this PR was trying to achieve.

The only reason RST_STREAM is handled is because I was concerned how to let streams decide whether they wanted to drain any buffered data if there was an error. GOAWAY is was very similar to RST_STREAM, but it also helped flesh out how non-stream frames would behave.